### PR TITLE
Add token-level transcription output with per-token log-probabilities

### PR DESCRIPTION
**Pull request by visheshd**
### Summary
Adds a token-level output variant to the Nemotron streaming decoder. The decode
loop already computes the winning logit for each emitted token and currently
discards it, returning only the joined string. This PR surfaces that information
through new methods without changing any existing behavior.

### Motivation
The current API returns only a `String`, so callers can't access information the
decoder has already produced internally. Exposing per-token data enables:

- confidence indicators / filtering low-confidence output in downstream UIs
- per-token inspection for evaluation and debugging
- any post-processing that needs to know which tokens the model was unsure about

This mirrors the existing `TimedToken` on the CTC/TDT path, bringing the Nemotron
path closer to parity.

### Changes
- New public struct `TokenInfo { id, text, logprob }` (same conventions as
  `TimedToken`: derives `Debug + Clone`).
- New method `transcribe_chunk_with_tokens(&mut self, &[f32]) -> Result<Vec<TokenInfo>>`
  (streaming), plus an offline companion
  `transcribe_audio_with_tokens(&mut self, &[f32]) -> Result<Vec<TokenInfo>>`.
- `logprob` is a numerically stable log-softmax over the full vocab logits at the
  chosen index. Blank tokens are skipped.
- Shared decode logic refactored into private helpers (`decode_chunk_tokens`,
  `process_chunk`); `transcribe_chunk` / `transcribe_audio` keep identical
  behavior and signatures (`transcribe_chunk` now builds its string by
  concatenating the per-token text pieces).

### Compatibility
Purely additive. No existing public signatures, behavior, or tests are modified.
No new dependencies (uses `std` for the log-softmax).

### Testing
- `cargo build`, `cargo clippy`, `cargo test` all clean.
- Added unit tests for the numerically stable log-softmax helper, asserting every
  value is finite and `<= 0.0`, that it peaks at the argmax index, and that the
  probabilities sum to 1. These follow the crate's existing convention of inline
  `#[cfg(test)]` tests over synthetic data (no model fixtures are committed).
- By construction, `transcribe_chunk`'s output equals the concatenation of the
  `text` fields from `transcribe_chunk_with_tokens` — both derive from the same
  decode path — so the exposed tokens are the decoder's actual decisions, not a
  separate approximation.

### CODE REVIEW

**Comment by altunenes** [comment]
hi sorry a bit of late.. I was working with this one https://github.com/altunenes/parakeet-rs/pull/104
Will check this soon. Thank you 😊

File: src/nemotron.rs [hunk 1]
```diff
@@ -35,6 +35,24 @@ const DECODER_LSTM_DIM: usize = 640;
 const CHUNK_SIZE: usize = 56;
 const PRE_ENCODE_CACHE: usize = 9;
 
+/// Token-level transcription output with its per-token log-probability.
+#[derive(Debug, Clone)]
+pub struct TokenInfo {
```
Comment (altunenes) [inline src/nemotron.rs:30]: 2 small thoughts on here:
`id: u32`... the rest of this file uses `usize` for token ids (see `decode_single(usize)`, `accumulated_tokens: Vec<usize>`). You're converting `usize → u32 → usize` round-trips. If it was a deliberate memory choice (TokenInfo Vecs can get long on long-form audio? ),  could you mention it in the field doc?

File: src/nemotron.rs [hunk 2]
```diff
@@ -35,6 +35,24 @@ const DECODER_LSTM_DIM: usize = 640;
 const CHUNK_SIZE: usize = 56;
 const PRE_ENCODE_CACHE: usize = 9;
 
+/// Token-level transcription output with its per-token log-probability.
+#[derive(Debug, Clone)]
```
Comment (altunenes) [inline src/nemotron.rs]: Worth deriving `PartialEq` here for testability downstream users will probably want to write assertions like `assert_eq!(tokens[0].id, 42)`. `Debug + Clone + PartialEq` is the standard "data struct" triple...

File: src/nemotron.rs [hunk 3]
```diff
@@ -35,6 +35,24 @@ const DECODER_LSTM_DIM: usize = 640;
...
+    /// Vocabulary id of the emitted (non-blank) token.
+    pub id: u32,
+    /// Decoded SentencePiece piece for this id.
+    pub text: String,
+    /// Log-softmax value of the chosen token.
+    pub logprob: f32,
+}
+
+/// Numerically stable log-softmax value at `idx` over the full logits vector.
+fn log_softmax_at(logits: &[f32], idx: usize) -> f32 {
```
Comment (altunenes) [inline src/nemotron.rs]: Tiny perf note for awareness: this iterates the full vocab twice (max, then exp-sum) per emitted token. For the multilingual model that's 1308 elements × 2 ≈ 26k float ops per token. Just calling out the cost in case anyone wonders later why we don't cache the LSE...

File: src/nemotron.rs [hunk 4]
```diff
@@ -556,7 +676,12 @@ impl Nemotron {
                     break;
                 }
 
-                tokens.push(max_idx);
+                let logprob = log_softmax_at(logits.as_slice().expect("contiguous logits"), max_idx);
```
Comment (altunenes) [inline src/nemotron.rs]: yeah in practice `Array1::from_vec(l_data.to_vec())` over in `model_nemotron.rs::run_decoder` always produces a contiguous standard layout array, so this *won't* panic probably and I proably do the same.. but maybe this better? 
    let logits_slice = logits
        .as_slice()   
        .ok_or_else(|| Error::Model("non-contiguous logits".into()))?;
    let logprob = log_softmax_at(logits_slice, max_idx);

File: src/nemotron.rs [hunk 5]
```diff
@@ -410,11 +428,103 @@ impl Nemotron {
...
+        }
+        Ok(result)
+    }
+
+    /// Stream transcribe a chunk of audio, returning token-level output with
+    /// per-token log-probabilities for the tokens emitted by this chunk.
+    ///
+    /// Additive companion to `transcribe_chunk`; shares identical buffering,
+    /// encoder/decoder, and state-update logic.
+    pub fn transcribe_chunk_with_tokens(
```
Comment (altunenes) [inline src/nemotron.rs:668]: After you rebase against master, this method will need to think about language tags. The multilingual Nemotron emits SentencePiece pieces like `<en-US>` inline with text; current `transcribe_chunk` strips them via `self.lang_tag_ids` filtering.
sorry for conflicts :( 
but what I suggest on here is always expose every token including `<lang>` tags?

File: src/nemotron.rs [hunk 6]
```diff
@@ -490,8 +600,9 @@ impl Nemotron {
         };
         self.encoder_cache = new_cache;
 
-        let tokens = self.decode_chunk(&encoded, enc_len as usize)?;
-        self.accumulated_tokens.extend(&tokens);
+        let tokens = self.decode_chunk_tokens(&encoded, enc_len as usize)?;
+        self.accumulated_tokens
```
Comment (altunenes) [inline src/nemotron.rs]: maybe something like: (after rebase?)
self.accumulated_tokens.extend(
        tokens.iter().map(|t| t.id as usize)
            .filter(|t| *t < self.vocab_size && !self.lang_tag_ids.contains(t))
    );

**Review by altunenes — CHANGES_REQUESTED** [review]
Sorry for merge conflicts but I would like to ask rebase.. made some suggestions also based on those but its cool! thank you 🙂

**Comment by visheshd** [comment]
@altunenes Thanks for the detailed review! I've rebased onto the current master (picking up the multilingual model from #104) and addressed all the points. Pushed as c39cac5.

  1. id: u32 round-tripping usize→u32→usize. Changed to usize to match the rest of the module (decode_single, accumulated_tokens, vocab ids), with a doc comment noting why.

  2. Derive PartialEq. Done: #[derive(Debug, Clone, PartialEq)] on TokenInfo.

  3. log-softmax walking the vocab twice. Good catch. log_softmax_at now takes the max_logit, the greedy decode loop already computes for the argmax, so it's a single exp-sum pass instead of a second max pass. Documented the reuse
  inline.

  4. .expect("contiguous logits"). Replaced with the non-panicking form:
  let logits_slice = logits
      .as_slice()
      .ok_or_else(|| Error::Model("non-contiguous logits".into()))?;

  5. Rebase + multilingual <lang> tags. Rebased. The *_with_tokens methods now expose every emitted token verbatim, including language-tag pieces like <en-US>, so callers can inspect or filter them. The string paths
  (transcribe_audio/transcribe_chunk) and the accumulated transcript keep stripping lang tags and out-of-vocab ids, so existing behavior is unchanged.

  To keep behavior identical, I refactored the decode into shared private helpers (process_audio/process_chunk/decode_chunk_tokens); the existing string-returning methods are now thin filter-and-join wrappers over them.

 **One deviation to flag**: on the accumulated_tokens filtering, your suggestion there was relative to the pre-rebase diff, so rather than filter on insert I kept accumulated_tokens unfiltered, matching the existing pattern where
  get_transcript strips lang tags and out-of-vocab ids at read time. This keeps the streaming accumulation behavior identical to what's on master today. 

Let me know if you'd like anything adjusted.

**Comment by altunenes** [comment]
Thank you very much for your contributions 😊
